### PR TITLE
fix: Fix reduce constraint bug

### LINKED ISSUE #31: [BUG] Reduction models symbolic inference does not work properly
### Describe the Bug
Playing with reduction models, in some particular cases symbolic shape inference does not work properly

### To Reproduce
```python
from mithril.models import *
model = Model()
model += Buffer()(IOKey(shape=(("Var1", ...), "a")))
model += Mean(axis=TBD, keepdim=True)(axis=-1)
print(model.shapes)
```
prints:
"{'$_Buffer_0_output': ['(V1, ...)', 'u1'], '$_Mean_1_output': ['(V1, ...)'], '$input': ['(V1, ...)', 'u1'], '$axis': None, '$keepdim': None}"

### Expected Behavior
"$_Mean_1_output" shape should be "['(V1, ...)', 1]"

### System Info
- **OS**: macOs 15.0
- **Mithril Version**: 0.1.1
- **Torch Version**: ~
- **MLX Version**: ~
- **NumPy Version**: ~
- **JAX Version**: ~
- **Python Version**: 3.12
- **Architecture**: Arm64

### PR DESCRIPTION
**Pull request by aturker-synnada**
## Description

Reduce constraint symbolic shape inference bug fixed. Closes #31 

## What is Changed
-  Reduce constraint updated.
- Tests for the bug added.

## Checklist:

- [x] Tests that cover the code added.
- [x] Corresponding changes documented.
- [x] All tests passed.
- [x] The code linted and styled (pre-commit run --all-files has passed).

### CODE REVIEW

File: mithril/framework/constraints.py [hunk 1]
```diff
@@ -1352,7 +1352,7 @@ def bcast_mat_mul_check(
 
 
 def reduce_constraints(
-    output: Tensor, input: Tensor, axis: Scalar, keepdim: Scalar | None = None
+    output: Tensor, input: Tensor, axis: Scalar, keepdim: Scalar
```
Comment (kberat-synnada) [inline mithril/framework/constraints.py:1355]: Do we need this?
'keepdim_val = keepdim.value if keepdim is not None else False'

File: mithril/framework/constraints.py [hunk 2]
```diff
@@ -1398,16 +1398,22 @@ def reduce_constraints(
...
-                        uni
-                        for idx, uni in enumerate(input_shape.prefix)
-                        if idx not in positive_axes
-                    ]
-                    repr_suffix: list[Uniadic] = [
-                        uni
-                        for idx, uni in enumerate(input_shape.reverse)
-                        if -(idx + 1) not in negative_axes
-                    ][::-1]
+                    repr_prefix: list[Uniadic] = []
```
Comment (kberat-synnada) [inline mithril/framework/constraints.py:1401]: we could operate only if keepdim is not TBD

File: tests/scripts/test_constraints.py [hunk 3]
```diff
@@ -1371,8 +1371,8 @@ def test_reduce_forward_2():
         "output": ["a", "b"],
         "input": [3, 4, 5, 6],
     }
-    final_shapes = {"output": [3, 5], "input": [3, 4, 5, 6], "axis": []}
-    scalar_info = {"axis": Scalar(value=(1, 3))}
+    final_shapes = {"output": [3, 5], "input": [3, 4, 5, 6], "axis": [], "keepdim": []}
+    scalar_info = {"axis": Scalar(value=(1, 3)), "keepdim": Scalar(value=False)}
```
Comment (kberat-synnada) [inline tests/scripts/test_constraints.py:1375]: we could add tests without given value to keepdim Scalar (e.g. {"keepdim": Scalar() })

File: mithril/framework/constraints.py [hunk 4]
```diff
@@ -1373,7 +1373,7 @@ def reduce_constraints(
         if isinstance(axis_val, int):
             axis_val = (axis_val,)
         elif axis_val is None:
-            if not keepdim_val:
+            if keepdim_val is False:
```
Comment (kberat-synnada) [inline mithril/framework/constraints.py:1376]: could we make any operations if keepdim is TBD, if not we could guard at the beginning ("if axis_val is not TBD and keepdim_val is not TBD")

Comment (aturker-synnada) [reply mithril/framework/constraints.py:1376]: We can extract input uniadics from axis_val even if keep dim is TBD

**Pull request merged**